File: context_runtime/compression/structural.py
"""Structural compression — a port of sidekick ``context_budget.clip`` (SPEC §5.5).

Lossy but provenance-preserving: keeps head+tail with an elision marker, records what
it was derived from and roughly what it omitted. The semantic compressor (LLMLingua-2)
is the v0.1 optional second stage; this structural pass is always available offline.
"""
from __future__ import annotations

from ..types import Compressed, Hit

_CHARS_PER_TOKEN = 4


def clip(text: str, max_chars: int = 4000) -> str:
    if not text:
        return ""
    if len(text) <= max_chars:
        return text
    head = max_chars * 2 // 3
    tail = max_chars - head
    omitted = len(text) - head - tail
    return f"{text[:head]}\n... [clipped {omitted} chars] ...\n{text[-tail:]}"


class StructuralCompressor:
    def compress(self, text: str, target_tokens: int) -> Compressed:
        max_chars = max(200, target_tokens * _CHARS_PER_TOKEN)
        clipped = clip(text, max_chars)
        return Compressed(
            text=clipped,
            tokens=max(1, len(clipped) // _CHARS_PER_TOKEN),
            omitted=("clipped-middle",) if len(clipped) < len(text) else (),
        )

    def assemble(self, hits: list[Hit], target_tokens: int) -> Compressed:
        """Pack ranked hits into a citation-numbered context within a token budget."""
        budget_chars = max(400, target_tokens * _CHARS_PER_TOKEN)
        parts: list[str] = []
        used = 0
        derived: list[str] = []
        omitted: list[str] = []
        for i, h in enumerate(hits, 1):
            block = f"[{i}] {h.filename}: {h.text}"
            if used + len(block) > budget_chars and parts:
                omitted.append(h.chunk_id)
                continue
            block = clip(block, budget_chars - used) if used + len(block) > budget_chars else block
            parts.append(block)
            derived.append(h.chunk_id)
            used += len(block)
        text = "\n\n".join(parts)
        return Compressed(
            text=text, tokens=max(1, len(text) // _CHARS_PER_TOKEN),
            derived_from=tuple(derived), omitted=tuple(omitted),
        )

File: go/compression/structural.go
package compression

import (
	"fmt"
	"strings"
	"unicode/utf8"

	"github.com/redevops-io/context-runtime/crtypes"
)

const charsPerToken = 4

// Compressed is the result of a structural compression pass.
type Compressed struct {
	Text         string   `json:"text"`
	Tokens       int      `json:"tokens"`
	DerivedFrom  []string `json:"derived_from"`
	Omitted      []string `json:"omitted"`
	RefreshAfter *string  `json:"refresh_after"`
}

// Clip preserves the head and tail of text around an elision marker when text is
// longer than maxChars. Character counts are based on UTF-8 runes.
func Clip(text string, maxChars int) string {
	if text == "" {
		return ""
	}

	runes := []rune(text)
	if maxChars < 0 {
		maxChars = 0
	}
	if len(runes) <= maxChars {
		return text
	}

	head := maxChars * 2 / 3
	tail := maxChars - head
	omitted := len(runes) - head - tail

	var tailText string
	if tail > 0 {
		tailText = string(runes[len(runes)-tail:])
	}

	return fmt.Sprintf("%s\n... [clipped %d chars] ...\n%s", string(runes[:head]), omitted, tailText)
}

// StructuralCompressor performs offline, provenance-preserving compression.
type StructuralCompressor struct{}

// Compress clips a single text blob to approximately targetTokens.
func (StructuralCompressor) Compress(text string, targetTokens int) Compressed {
	maxChars := maxInt(200, targetTokens*charsPerToken)
	clipped := Clip(text, maxChars)
	omitted := []string(nil)
	if utf8.RuneCountInString(clipped) < utf8.RuneCountInString(text) {
		omitted = []string{"clipped-middle"}
	}

	return Compressed{
		Text:    clipped,
		Tokens:  maxInt(1, utf8.RuneCountInString(clipped)/charsPerToken),
		Omitted: omitted,
	}
}

// Assemble packs ranked hits into a citation-numbered context within a token
// budget. Included hits are cited as [1], [2], ... using their original rank.
func (StructuralCompressor) Assemble(hits []crtypes.Hit, targetTokens int) Compressed {
	budgetChars := maxInt(400, targetTokens*charsPerToken)
	parts := make([]string, 0, len(hits))
	derived := make([]string, 0, len(hits))
	omitted := []string(nil)
	used := 0

	for i, h := range hits {
		block := fmt.Sprintf("[%d] %s: %s", i+1, h.Filename, h.Text)
		blockLen := utf8.RuneCountInString(block)
		if used+blockLen > budgetChars && len(parts) > 0 {
			omitted = append(omitted, h.ChunkID)
			continue
		}
		if used+blockLen > budgetChars {
			block = Clip(block, budgetChars-used)
			blockLen = utf8.RuneCountInString(block)
		}
		parts = append(parts, block)
		derived = append(derived, h.ChunkID)
		used += blockLen
	}

	text := strings.Join(parts, "\n\n")
	return Compressed{
		Text:        text,
		Tokens:      maxInt(1, utf8.RuneCountInString(text)/charsPerToken),
		DerivedFrom: derived,
		Omitted:     omitted,
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
